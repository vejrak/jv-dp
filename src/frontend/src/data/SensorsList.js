// @flow

import React from 'react'
import { Button } from 'react-bootstrap'
import { getSensors } from '../sensor/actions'
import type { ErrorResponse, Sensor } from '../types'
import Spinner from '../components/Spinner'
import Input from '../components/Input'
import { filterSensors } from './helpers'

type Props = $ReadOnly<{|
  getSensors: typeof getSensors,
  isFetching: boolean,
  error: ?ErrorResponse,
  sensors: Array<Sensor>,
  showDetail: Function,
|}>

type State = {
  sensorSearch: string,
}

class SensorsList extends React.PureComponent<Props, State> {
  state = {
    sensorSearch: '',
  }

  componentDidMount() {
    this.props.getSensors()
  }

  handleChange = (sensorSearch: Object) => {
    this.setState(sensorSearch)
  }

  render() {
    const { isFetching, error, sensors, showDetail } = this.props
    const { sensorSearch } = this.state
    const filtredSensors = filterSensors(sensors, sensorSearch)
    if (isFetching) return <Spinner />

    if (error)
      return (
        <div className="alert alert-danger" role="alert">
          {error.errmsg}
        </div>
      )

    if (!sensors || sensors.length === 0) return null
    return (
      <>
        <h2>Sensors</h2>
        <label class-name="mt-2" htmlFor="sensorSearch">
          Search sensor
        </label>
        <Input
          className="form-control mb-3"
          id="sensorSearch"
          name="sensorSearch"
          onChange={({ target }) =>
            this.handleChange({ sensorSearch: target.value })
          }
          value={sensorSearch}
        />
        {filtredSensors.length > 0 && (
          <table className="table">
            <thead>
              <tr>
                <th scope="col">Identificator</th>
                <th scope="col">Description</th>
                <th />
              </tr>
            </thead>
            {filtredSensors.map((sensor) => (
              <tbody key={sensor._id} className="mb-4">
                <tr>
                  <td>{sensor.identificator}</td>
                  <td>{sensor.description}</td>
                  <td>
                    <Button onClick={() => showDetail(sensor._id)}>
                      {' '}
                      Show{' '}
                    </Button>
                  </td>
                </tr>
              </tbody>
            ))}
          </table>
        )}
      </>
    )
  }
}

export default SensorsList
