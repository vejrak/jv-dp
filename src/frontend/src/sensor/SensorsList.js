// @flow

import React from 'react'
import { Button } from 'react-bootstrap'
import filter from 'lodash/filter'
import { deleteSensor, editSensor, getSensors } from './actions'
import { getGroups } from '../group/actions'
import { canShowUpdateForm } from '../helpers'
import SensorForm from './form'
import type { ErrorResponse, Sensor } from '../types'
import Spinner from '../components/Spinner'

type Props = $ReadOnly<{|
  deleteSensor: typeof deleteSensor,
  getSensors: typeof getSensors,
  getGroups: typeof getGroups,
  editSensor: typeof editSensor,
  isFetching: boolean,
  error: ?ErrorResponse,
  sensors: Array<Sensor>,
|}>

type State = {
  showEditFormsFor: Array<string>,
}

class SensorsList extends React.PureComponent<Props, State> {
  state = {
    showEditFormsFor: [],
  }

  componentDidMount() {
    this.props.getSensors()
    this.props.getGroups()
  }

  handleEditClick = (_id: string): void => {
    const { showEditFormsFor } = this.state
    this.setState({ showEditFormsFor: showEditFormsFor.concat(_id) })
  }

  handleHideClick = (_id: string): void => {
    const { showEditFormsFor } = this.state
    this.setState({
      showEditFormsFor: filter(showEditFormsFor, _id),
    })
  }

  updateSensor = (_id: string): Function => (sensor: Sensor): void => {
    this.props.editSensor(sensor, _id)
  }

  handleDeleteClick = async (_id: string) => {
    await this.props.deleteSensor(_id)
  }

  render() {
    const { showEditFormsFor } = this.state
    const { isFetching, error, sensors } = this.props

    if (isFetching) return <Spinner />

    if (error)
      return (
        <div className="alert alert-danger" role="alert">
          {error.errmsg}
        </div>
      )

    if (!sensors || sensors.length === 0) return null
    return (
      <table className="table">
        <thead>
          <tr>
            <th scope="col">Identificator</th>
            <th scope="col">Description</th>
            <th scope="col">Datasource</th>
            <th scope="col">Separator</th>
            <th />
            <th />
          </tr>
        </thead>
        {sensors.map((sensor) => (
          <tbody key={sensor._id} className="mb-4">
            <tr>
              <td>{sensor.identificator}</td>
              <td>{sensor.description}</td>
              <td>{sensor.datasource_type}</td>
              <td>{sensor.separator}</td>
              <td>
                <Button
                  className="mt-3 mb-3 btn d-block btn-danger"
                  onClick={() => {
                    if (
                      window.confirm(
                        'Are you sure you wish to delete this item?',
                      )
                    )
                      this.handleDeleteClick(sensor._id)
                  }}
                >
                  Delete
                </Button>
              </td>
              {!canShowUpdateForm(showEditFormsFor, sensor._id) && (
                <td>
                  <Button
                    className="mt-3 mb-3"
                    onClick={() => this.handleEditClick(sensor._id)}
                  >
                    Edit
                  </Button>
                </td>
              )}
            </tr>
            {canShowUpdateForm(showEditFormsFor, sensor._id) && (
              <>
                <tr>
                  <td colSpan="4">
                    <SensorForm
                      sensor={sensor}
                      updateSensor={this.updateSensor(sensor._id)}
                    />
                  </td>
                  <td>
                    <Button
                      className="mt-3 mb-3"
                      onClick={() => this.handleHideClick(sensor._id)}
                    >
                      Hide
                    </Button>
                  </td>
                </tr>
              </>
            )}
          </tbody>
        ))}
      </table>
    )
  }
}

export default SensorsList
