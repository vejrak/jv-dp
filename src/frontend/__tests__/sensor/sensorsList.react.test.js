import React from 'react'
import renderer from 'react-test-renderer'
import SensorsList from '../../src/sensor/SensorsList'

const mockProps = {
  deleteSensor: () => {},
  editSensor: () => {},
  getSensors: () => {},
  getGroups: () => {},
  isFetching: false,
  error: null,
  sensors: [{ _id: 'id1', identificator: 'id1', description: 'test' }],
}

it('renders correctly', () => {
  const tree = renderer.create(<SensorsList {...mockProps} />).toJSON()
  expect(tree).toMatchSnapshot()
})
