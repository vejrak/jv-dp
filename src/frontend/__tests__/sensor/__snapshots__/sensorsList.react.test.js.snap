// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<table
  className="table"
>
  <thead>
    <tr>
      <th
        scope="col"
      >
        Identificator
      </th>
      <th
        scope="col"
      >
        Description
      </th>
      <th
        scope="col"
      >
        Datasource
      </th>
      <th
        scope="col"
      >
        Separator
      </th>
      <th />
      <th />
    </tr>
  </thead>
  <tbody
    className="mb-4"
  >
    <tr>
      <td>
        id1
      </td>
      <td>
        test
      </td>
      <td />
      <td />
      <td>
        <button
          className="mt-3 mb-3 btn d-block btn-danger btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Delete
        </button>
      </td>
      <td>
        <button
          className="mt-3 mb-3 btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Edit
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;
